File: routers/bookmarks.py
"""
routers/bookmarks.py — bookmark management and code assignment.

  POST   /api/bookmarks
  GET    /api/bookmarks
  GET    /api/bookmarks/ids
  DELETE /api/bookmarks/{bookmark_id}
  DELETE /api/bookmarks/by-msg/{msg_id}
  POST   /api/bookmarks/{bookmark_id}/codes/{code_id}
  DELETE /api/bookmarks/{bookmark_id}/codes/{code_id}

  Back-compat aliases (old /labels/ paths):
  POST   /api/bookmarks/{bookmark_id}/labels/{label_id}
  DELETE /api/bookmarks/{bookmark_id}/labels/{label_id}
"""

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, HTTPException, Request

import state
from database import get_db, get_session_user

router = APIRouter()


def _uid(request: Request) -> Optional[int]:
    """Return the authenticated user's id in multi mode, else None."""
    if state.app_mode != "multi":
        return None
    token = request.cookies.get("session")
    if not token:
        return None
    user = get_session_user(token)
    return user["id"] if user else None


@router.post("/api/bookmarks")
async def add_bookmark(body: dict, request: Request):
    msg_id = body.get("msg_id")
    if not isinstance(msg_id, int):
        raise HTTPException(400, "msg_id (int) is required")
    ctx_before = int(body.get("ctx_before", 5))
    ctx_after  = int(body.get("ctx_after",  5))
    note       = str(body.get("note", ""))
    user_id    = _uid(request)
    conn = get_db()
    try:
        row = conn.execute("SELECT id FROM messages WHERE id=?", (msg_id,)).fetchone()
        if not row:
            raise HTTPException(404, "Message not found")
        if user_id is not None:
            existing = conn.execute(
                "SELECT id FROM bookmarks WHERE msg_id=? AND user_id=?", (msg_id, user_id)
            ).fetchone()
        else:
            existing = conn.execute(
                "SELECT id FROM bookmarks WHERE msg_id=?", (msg_id,)
            ).fetchone()
        if existing:
            return {"status": "exists", "bookmark_id": existing["id"]}
        cur = conn.execute(
            "INSERT INTO bookmarks (msg_id, ctx_before, ctx_after, note, created_at, user_id) VALUES (?,?,?,?,?,?)",
            (msg_id, ctx_before, ctx_after, note, datetime.utcnow().isoformat(), user_id),
        )
        conn.commit()
        return {"status": "created", "bookmark_id": cur.lastrowid}
    finally:
        conn.close()


@router.get("/api/bookmarks")
async def list_bookmarks(request: Request):
    user_id = _uid(request)
    conn = get_db()
    if user_id is not None:
        rows = conn.execute(
            """SELECT b.id AS bookmark_id, b.ctx_before, b.ctx_after, b.note, b.created_at,
                      m.*
               FROM bookmarks b
               JOIN messages m ON m.id = b.msg_id
               WHERE b.user_id = ?
               ORDER BY b.created_at DESC""",
            (user_id,),
        ).fetchall()
        bm_ids_sub = "SELECT id FROM bookmarks WHERE user_id = ?"
        code_rows = conn.execute(
            f"""SELECT bc.bookmark_id, c.id, c.name, c.color, c.description, c.category_id
               FROM bookmark_codes bc
               JOIN codes c ON c.id = bc.code_id
               WHERE bc.bookmark_id IN ({bm_ids_sub})""",
            (user_id,),
        ).fetchall()
        hl_rows = conn.execute(
            f"""SELECT bch.id, bch.bookmark_id, bch.code_id, bch.highlighted_text,
                       c.name AS code_name, c.color AS code_color
               FROM bookmark_code_highlights bch
               JOIN codes c ON c.id = bch.code_id
               WHERE bch.bookmark_id IN ({bm_ids_sub})""",
            (user_id,),
        ).fetchall()
    else:
        rows = conn.execute(
            """SELECT b.id AS bookmark_id, b.ctx_before, b.ctx_after, b.note, b.created_at,
                      m.*
               FROM bookmarks b
               JOIN messages m ON m.id = b.msg_id
               ORDER BY b.created_at DESC"""
        ).fetchall()
        code_rows = conn.execute(
            """SELECT bc.bookmark_id, c.id, c.name, c.color, c.description, c.category_id
               FROM bookmark_codes bc
               JOIN codes c ON c.id = bc.code_id"""
        ).fetchall()
        hl_rows = conn.execute(
            """SELECT bch.id, bch.bookmark_id, bch.code_id, bch.highlighted_text,
                      c.name AS code_name, c.color AS code_color
               FROM bookmark_code_highlights bch
               JOIN codes c ON c.id = bch.code_id"""
        ).fetchall()
    conn.close()

    codes_by_bm: dict = {}
    for cr in code_rows:
        codes_by_bm.setdefault(cr["bookmark_id"], []).append({
            "id": cr["id"], "name": cr["name"], "color": cr["color"],
            "description": cr["description"], "category_id": cr["category_id"],
        })

    hl_by_bm: dict = {}
    for hr in hl_rows:
        hl_by_bm.setdefault(hr["bookmark_id"], []).append({
            "id": hr["id"], "code_id": hr["code_id"],
            "code_name": hr["code_name"], "code_color": hr["code_color"],
            "highlighted_text": hr["highlighted_text"],
        })

    result = []
    for r in rows:
        d = dict(r)
        d["codes"]      = codes_by_bm.get(d["bookmark_id"], [])
        d["labels"]     = d["codes"]   # back-compat alias
        d["highlights"] = hl_by_bm.get(d["bookmark_id"], [])
        result.append(d)
    return result


@router.get("/api/bookmarks/ids")
async def list_bookmark_ids(request: Request):
    user_id = _uid(request)
    conn = get_db()
    if user_id is not None:
        rows = conn.execute(
            "SELECT msg_id FROM bookmarks WHERE user_id = ?", (user_id,)
        ).fetchall()
    else:
        rows = conn.execute("SELECT msg_id FROM bookmarks").fetchall()
    conn.close()
    return [r["msg_id"] for r in rows]


@router.patch("/api/bookmarks/{bookmark_id}")
async def update_bookmark(bookmark_id: int, body: dict, request: Request):
    note = body.get("note")
    if note is None:
        raise HTTPException(400, "note is required")
    user_id = _uid(request)
    conn = get_db()
    if user_id is not None:
        cur = conn.execute(
            "UPDATE bookmarks SET note = ? WHERE id = ? AND user_id = ?",
            (str(note), bookmark_id, user_id),
        )
    else:
        cur = conn.execute(
            "UPDATE bookmarks SET note = ? WHERE id = ?",
            (str(note), bookmark_id),
        )
    conn.commit()
    conn.close()
    if cur.rowcount == 0:
        raise HTTPException(404, "Bookmark not found")
    return {"status": "updated"}


@router.delete("/api/bookmarks/{bookmark_id}")
async def delete_bookmark(bookmark_id: int, request: Request):
    user_id = _uid(request)
    conn = get_db()
    if user_id is not None:
        cur = conn.execute(
            "DELETE FROM bookmarks WHERE id=? AND user_id=?", (bookmark_id, user_id)
        )
    else:
        cur = conn.execute("DELETE FROM bookmarks WHERE id=?", (bookmark_id,))
    conn.commit()
    conn.close()
    if cur.rowcount == 0:
        raise HTTPException(404, "Bookmark not found")
    return {"status": "deleted"}


@router.delete("/api/bookmarks/by-msg/{msg_id}")
async def delete_bookmark_by_msg(msg_id: int, request: Request):
    user_id = _uid(request)
    conn = get_db()
    if user_id is not None:
        cur = conn.execute(
            "DELETE FROM bookmarks WHERE msg_id=? AND user_id=?", (msg_id, user_id)
        )
    else:
        cur = conn.execute("DELETE FROM bookmarks WHERE msg_id=?", (msg_id,))
    conn.commit()
    conn.close()
    return {"status": "deleted", "affected": cur.rowcount}


# ── Code assignment ───────────────────────────────────────────────────────────

@router.post("/api/bookmarks/{bookmark_id}/codes/{code_id}")
async def assign_code(bookmark_id: int, code_id: int):
    conn = get_db()
    try:
        conn.execute(
            "INSERT OR IGNORE INTO bookmark_codes (bookmark_id, code_id) VALUES (?,?)",
            (bookmark_id, code_id),
        )
        conn.commit()
        return {"status": "assigned"}
    finally:
        conn.close()


@router.delete("/api/bookmarks/{bookmark_id}/codes/{code_id}")
async def unassign_code(bookmark_id: int, code_id: int):
    conn = get_db()
    conn.execute(
        "DELETE FROM bookmark_codes WHERE bookmark_id=? AND code_id=?",
        (bookmark_id, code_id),
    )
    conn.commit()
    conn.close()
    return {"status": "unassigned"}


# ── Back-compat: old /labels/ paths ──────────────────────────────────────────

@router.post("/api/bookmarks/{bookmark_id}/labels/{label_id}")
async def assign_label_compat(bookmark_id: int, label_id: int):
    return await assign_code(bookmark_id, label_id)

@router.delete("/api/bookmarks/{bookmark_id}/labels/{label_id}")
async def unassign_label_compat(bookmark_id: int, label_id: int):
    return await unassign_code(bookmark_id, label_id)


# ── Span-level highlight codings ─────────────────────────────────────────────

@router.post("/api/bookmarks/{bookmark_id}/highlights")
async def add_highlight(bookmark_id: int, body: dict):
    code_id          = body.get("code_id")
    highlighted_text = (body.get("highlighted_text") or "").strip()
    if not isinstance(code_id, int) or not highlighted_text:
        raise HTTPException(400, "code_id (int) and highlighted_text (str) are required")
    conn = get_db()
    try:
        cur = conn.execute(
            """INSERT OR IGNORE INTO bookmark_code_highlights
               (bookmark_id, code_id, highlighted_text, created_at)
               VALUES (?,?,?,?)""",
            (bookmark_id, code_id, highlighted_text, datetime.utcnow().isoformat()),
        )
        conn.commit()
        # If INSERT OR IGNORE skipped (duplicate), fetch the existing row's id
        if cur.lastrowid and cur.rowcount > 0:
            return {"status": "created", "id": cur.lastrowid}
        existing = conn.execute(
            """SELECT id FROM bookmark_code_highlights
               WHERE bookmark_id=? AND code_id=? AND highlighted_text=?""",
            (bookmark_id, code_id, highlighted_text),
        ).fetchone()
        return {"status": "exists", "id": existing["id"] if existing else None}
    finally:
        conn.close()


@router.delete("/api/bookmarks/{bookmark_id}/highlights/{highlight_id}")
async def remove_highlight(bookmark_id: int, highlight_id: int):
    conn = get_db()
    conn.execute(
        "DELETE FROM bookmark_code_highlights WHERE id=? AND bookmark_id=?",
        (highlight_id, bookmark_id),
    )
    conn.commit()
    conn.close()
    return {"status": "deleted"}

File: backend/src/routes/bookmarks.rs
use axum::{
    extract::{Path, State},
    Json,
};
use serde_json::{json, Value};
use crate::{
    error::{AppError, Result},
    models::{AuthUser, CreateBookmarkRequest},
    state::AppState,
};

pub async fn create_bookmark(
    State(state): State<AppState>,
    user: AuthUser,
    Json(req): Json<CreateBookmarkRequest>,
) -> Result<Json<Value>> {
    let mode = state.get_app_mode().await;
    let user_id = if mode.as_deref() == Some("multi") {
        Some(user.id)
    } else {
        None
    };

    let ctx_before = req.ctx_before.unwrap_or(5);
    let ctx_after = req.ctx_after.unwrap_or(5);

    let id: i64 = sqlx::query_scalar(
        "INSERT INTO bookmarks (msg_id, ctx_before, ctx_after, note, user_id)
         VALUES ($1, $2, $3, $4, $5) RETURNING id",
    )
    .bind(req.msg_id)
    .bind(ctx_before)
    .bind(ctx_after)
    .bind(&req.note)
    .bind(user_id)
    .fetch_one(&state.db)
    .await?;

    Ok(Json(json!({ "id": id, "msg_id": req.msg_id })))
}

pub async fn list_bookmarks(
    State(state): State<AppState>,
    user: AuthUser,
) -> Result<Json<Value>> {
    let mode = state.get_app_mode().await;
    let bookmark_rows = if mode.as_deref() == Some("multi") {
        sqlx::query(
            "SELECT b.id, b.msg_id, b.ctx_before, b.ctx_after, b.note, b.created_at,
                    m.content, m.username, m.date, m.is_suno_team
             FROM bookmarks b
             JOIN messages m ON m.id = b.msg_id
             WHERE b.user_id = $1
             ORDER BY b.created_at DESC",
        )
        .bind(user.id)
        .fetch_all(&state.db)
        .await?
    } else {
        sqlx::query(
            "SELECT b.id, b.msg_id, b.ctx_before, b.ctx_after, b.note, b.created_at,
                    m.content, m.username, m.date, m.is_suno_team
             FROM bookmarks b
             JOIN messages m ON m.id = b.msg_id
             ORDER BY b.created_at DESC",
        )
        .fetch_all(&state.db)
        .await?
    };

    use sqlx::Row;
    use std::collections::HashMap;

    if bookmark_rows.is_empty() {
        return Ok(Json(json!([])));
    }

    // Collect IDs for batch sub-queries (our own DB integers — safe for interpolation)
    let bookmark_ids: Vec<i64> = bookmark_rows.iter().map(|r| r.get::<i64, _>("id")).collect();
    let id_list = bookmark_ids
        .iter()
        .map(|id| id.to_string())
        .collect::<Vec<_>>()
        .join(",");

    let codes_sql = format!(
        "SELECT bc.bookmark_id, c.id, c.name, c.color
         FROM bookmark_codes bc JOIN codes c ON c.id = bc.code_id
         WHERE bc.bookmark_id IN ({id})
         UNION
         SELECT bch.bookmark_id, c.id, c.name, c.color
         FROM bookmark_code_highlights bch JOIN codes c ON c.id = bch.code_id
         WHERE bch.bookmark_id IN ({id})",
        id = id_list
    );
    let codes_rows = sqlx::query(&codes_sql).fetch_all(&state.db).await?;

    let mut codes_map: HashMap<i64, Vec<Value>> = HashMap::new();
    for r in &codes_rows {
        let bm_id: i64 = r.get("bookmark_id");
        let code_id: i64 = r.get("id");
        let entry = codes_map.entry(bm_id).or_default();
        if !entry.iter().any(|c| c["id"].as_i64() == Some(code_id)) {
            entry.push(json!({
                "id": code_id,
                "name": r.get::<String, _>("name"),
                "color": r.get::<String, _>("color"),
            }));
        }
    }

    let hl_sql = format!(
        "SELECT bch.id, bch.bookmark_id, bch.code_id,
                c.name as code_name, c.color as code_color, bch.highlighted_text
         FROM bookmark_code_highlights bch
         JOIN codes c ON c.id = bch.code_id
         WHERE bch.bookmark_id IN ({})",
        id_list
    );
    let hl_rows = sqlx::query(&hl_sql).fetch_all(&state.db).await?;

    let mut hl_map: HashMap<i64, Vec<Value>> = HashMap::new();
    for r in &hl_rows {
        let bm_id: i64 = r.get("bookmark_id");
        hl_map.entry(bm_id).or_default().push(json!({
            "id": r.get::<i64, _>("id"),
            "code_id": r.get::<i64, _>("code_id"),
            "code_name": r.get::<String, _>("code_name"),
            "code_color": r.get::<String, _>("code_color"),
            "highlighted_text": r.get::<String, _>("highlighted_text"),
        }));
    }

    let bookmarks: Vec<Value> = bookmark_rows
        .iter()
        .map(|row| {
            let bm_id: i64 = row.get("id");
            json!({
                "id": bm_id,
                "msg_id": row.get::<i64, _>("msg_id"),
                "ctx_before": row.get::<i64, _>("ctx_before"),
                "ctx_after": row.get::<i64, _>("ctx_after"),
                "note": row.get::<Option<String>, _>("note"),
                "created_at": row.get::<String, _>("created_at"),
                "content": row.get::<String, _>("content"),
                "username": row.get::<String, _>("username"),
                "date": row.get::<String, _>("date"),
                "is_suno_team": row.get::<Option<String>, _>("is_suno_team"),
                "codes": codes_map.get(&bm_id).cloned().unwrap_or_default(),
                "highlights": hl_map.get(&bm_id).cloned().unwrap_or_default(),
            })
        })
        .collect();

    Ok(Json(json!(bookmarks)))
}

/// Lightweight bookmark list — only message metadata, no codes or highlights sub-queries.
pub async fn list_bookmarks_meta(
    State(state): State<AppState>,
    user: AuthUser,
) -> Result<Json<Value>> {
    let mode = state.get_app_mode().await;
    let rows = if mode.as_deref() == Some("multi") {
        sqlx::query(
            "SELECT b.id, b.msg_id, b.note, b.created_at,
                    m.content, m.username, m.date, m.is_suno_team
             FROM bookmarks b
             JOIN messages m ON m.id = b.msg_id
             WHERE b.user_id = $1
             ORDER BY b.created_at DESC",
        )
        .bind(user.id)
        .fetch_all(&state.db)
        .await?
    } else {
        sqlx::query(
            "SELECT b.id, b.msg_id, b.note, b.created_at,
                    m.content, m.username, m.date, m.is_suno_team
             FROM bookmarks b
             JOIN messages m ON m.id = b.msg_id
             ORDER BY b.created_at DESC",
        )
        .fetch_all(&state.db)
        .await?
    };

    use sqlx::Row;
    let bookmarks: Vec<Value> = rows
        .into_iter()
        .map(|r| json!({
            "id": r.get::<i64, _>("id"),
            "msg_id": r.get::<i64, _>("msg_id"),
            "note": r.get::<Option<String>, _>("note"),
            "created_at": r.get::<String, _>("created_at"),
            "content": r.get::<String, _>("content"),
            "username": r.get::<String, _>("username"),
            "date": r.get::<String, _>("date"),
            "is_suno_team": r.get::<Option<String>, _>("is_suno_team"),
        }))
        .collect();

    Ok(Json(json!(bookmarks)))
}

pub async fn list_bookmark_ids(
    State(state): State<AppState>,
    user: AuthUser,
) -> Result<Json<Value>> {
    let mode = state.get_app_mode().await;
    let ids: Vec<i64> = if mode.as_deref() == Some("multi") {
        sqlx::query_scalar("SELECT msg_id FROM bookmarks WHERE user_id = $1")
            .bind(user.id)
            .fetch_all(&state.db)
            .await?
    } else {
        sqlx::query_scalar("SELECT msg_id FROM bookmarks")
            .fetch_all(&state.db)
            .await?
    };

    Ok(Json(json!(ids)))
}

pub async fn delete_bookmark(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<Value>> {
    let affected = sqlx::query("DELETE FROM bookmarks WHERE id = $1")
        .bind(id)
        .execute(&state.db)
        .await?
        .rows_affected();

    if affected == 0 {
        return Err(AppError::NotFound(format!("Bookmark {} not found", id)));
    }
    Ok(Json(json!({ "deleted": id })))
}

pub async fn delete_bookmark_by_msg(
    State(state): State<AppState>,
    Path(msg_id): Path<i64>,
) -> Result<Json<Value>> {
    let affected = sqlx::query("DELETE FROM bookmarks WHERE msg_id = $1")
        .bind(msg_id)
        .execute(&state.db)
        .await?
        .rows_affected();

    Ok(Json(json!({ "deleted": affected, "msg_id": msg_id })))
}

pub async fn add_label(
    State(state): State<AppState>,
    Path((bookmark_id, label_id)): Path<(i64, i64)>,
) -> Result<Json<Value>> {
    sqlx::query(
        "INSERT INTO bookmark_labels (bookmark_id, label_id) VALUES ($1, $2)
         ON CONFLICT DO NOTHING",
    )
    .bind(bookmark_id)
    .bind(label_id)
    .execute(&state.db)
    .await?;

    Ok(Json(json!({ "bookmark_id": bookmark_id, "label_id": label_id })))
}

pub async fn remove_label(
    State(state): State<AppState>,
    Path((bookmark_id, label_id)): Path<(i64, i64)>,
) -> Result<Json<Value>> {
    sqlx::query(
        "DELETE FROM bookmark_labels WHERE bookmark_id = $1 AND label_id = $2",
    )
    .bind(bookmark_id)
    .bind(label_id)
    .execute(&state.db)
    .await?;

    Ok(Json(json!({ "removed": true })))
}

pub async fn add_highlight(
    State(state): State<AppState>,
    Path(bookmark_id): Path<i64>,
    Json(req): Json<crate::models::AddHighlightRequest>,
) -> Result<Json<Value>> {
    let id: i64 = sqlx::query_scalar(
        "INSERT INTO bookmark_code_highlights
         (bookmark_id, code_id, highlighted_text)
         VALUES ($1, $2, $3) RETURNING id",
    )
    .bind(bookmark_id)
    .bind(req.code_id)
    .bind(&req.highlighted_text)
    .fetch_one(&state.db)
    .await?;

    Ok(Json(json!({
        "id": id,
        "bookmark_id": bookmark_id,
        "code_id": req.code_id,
        "highlighted_text": req.highlighted_text,
    })))
}

pub async fn remove_highlight(
    State(state): State<AppState>,
    Path((bookmark_id, hl_id)): Path<(i64, i64)>,
) -> Result<Json<Value>> {
    sqlx::query(
        "DELETE FROM bookmark_code_highlights WHERE id = $1 AND bookmark_id = $2",
    )
    .bind(hl_id)
    .bind(bookmark_id)
    .execute(&state.db)
    .await?;
    Ok(Json(json!({ "deleted": hl_id })))
}
